crates/qasm2/src/parse.rs

```diff
@@ -0,0 +1,1812 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+//! The core of the parsing algorithm.  This module contains the core logic for the
+//! recursive-descent parser, which handles all statements of OpenQASM 2.  In places where we have
+//! to evaluate a mathematical expression on parameters, we instead swap to a short-lived
+//! operator-precedence parser.
+
+use hashbrown::{HashMap, HashSet};
+use pyo3::prelude::{PyObject, PyResult, Python};
+
+use crate::bytecode::InternalBytecode;
+use crate::error::{
+    message_bad_eof, message_generic, message_incorrect_requirement, Position, QASM2ParseError,
+};
+use crate::expr::{Expr, ExprParser};
+use crate::lex::{Token, TokenContext, TokenStream, TokenType, Version};
+use crate::{CustomClassical, CustomInstruction};
+
+/// The number of gates that are built in to the OpenQASM 2 language.  This is U and CX.
+const N_BUILTIN_GATES: usize = 2;
+/// The "qelib1.inc" special include.  The elements of the tuple are the gate name, the number of
+/// parameters it takes, and the number of qubits it acts on.
+const QELIB1: [(&str, usize, usize); 23] = [
+    ("u3", 3, 1),
+    ("u2", 2, 1),
+    ("u1", 1, 1),
+    ("cx", 0, 2),
+    ("id", 0, 1),
+    ("x", 0, 1),
+    ("y", 0, 1),
+    ("z", 0, 1),
+    ("h", 0, 1),
+    ("s", 0, 1),
+    ("sdg", 0, 1),
+    ("t", 0, 1),
+    ("tdg", 0, 1),
+    ("rx", 1, 1),
+    ("ry", 1, 1),
+    ("rz", 1, 1),
+    ("cz", 0, 2),
+    ("cy", 0, 2),
+    ("ch", 0, 2),
+    ("ccx", 0, 3),
+    ("crz", 1, 2),
+    ("cu1", 1, 2),
+    ("cu3", 3, 2),
+];
+
+lazy_static! {
+    static ref BUILTIN_CLASSICAL: HashSet<&'static str> = {
+        let mut out = HashSet::with_capacity(6);
+        out.insert("cos");
+        out.insert("exp");
+        out.insert("ln");
+        out.insert("sin");
+        out.insert("sqrt");
+        out.insert("tan");
+        out
+    };
+}
```

Comment:
Just a random thought instead of doing a lazy static hash set like this I wonder what the performance of just using a static array here is. You lose the constant time lookup, but even though checking if it's in a array is `O(n)` `n` is always 6. At the end of the day it's not a huge deal and far from likely to be a bottleneck, it's more just idle curiosity (and a small desire to remove an external dependency if we can).

Reply:
Removed in 5f74f1e04.